eip-0031.md

```diff
@@ -0,0 +1,340 @@
+
+Strong advantage of this approach (compared to Cardano’s one) is that user always knows in advance whether there is an opportunity to pay via “babel fees” or not, and if there is, what is the exchange ratio. He can therefore be (almost) certain that if he decides to use it, his transaction will be included in the blockchain ledger. Be aware, however, that there exist some exceptions to this rule, which is later discussed in the “Wallets implementation” section.
+
+Motivation
+--------------------------------------------
+
+Many users use blockchain solely for transferring (native)tokens, such as stablecoins, or even “meme” coins. These users, understandably, do not want to be bothered with keeping an eye on the amount of blockchain’s primary token they own, or even obtaining this primary token in the first place.
+
+Once they run out of primary token, they have to difficultly (and often costly), swap their tokens of interest for the primary tokens, that they can later use to cover transaction fees. Since primary tokens are also needed for these swaps, users may be forced to introduce new capital to their portfolio solely for the purpose of purchasing primary tokens, used for fee paying. 
+
+Since basic transactional fees on the Ergo blockchain are generally quite low, the "babel fees" users would be probaby willing to pay a fee that could be higher than that of a primary token transaction, in exchange for being able to pay in their token of interest and not having to bother with the blockchain’s primary token purchase.
+
+This brings up a financial incentive for “EIP-0031 supporters”, who could benefit out of this arbitrage by providing the liquidity for such “babel fees” users, with primary token’s selling price (expressed in tokens of interest) being higher compared to the same token pair on the exchanges.
+
+Smartcontract specification
+--------------------------------------------
+Here is the smartcontract's source code in ErgoScript which will be used to protect the babel fee box:
+```scala
+{
+
+    val babelFeeBoxCreator = SELF.R4[SigmaProp].get
+
+    val recreatedBox = OUTPUTS(OUTPUTS.size - 2)
+    
+    val babelFeeBoxRecreated = (
+        recreatedBox.propositionBytes == SELF.propositionBytes &&
+        recreatedBox.R4[SigmaProp].get == SELF.R4[SigmaProp].get &&
+        recreatedBox.R5[Long].get == SELF.R5[Long].get &&
+        recreatedBox.tokens(0)._1 == tokenId
+    )
+
+    val nanoErgsDifference = SELF.value - recreatedBox.value
+    val babelTokensBefore = if(SELF.tokens.size > 0){ SELF.tokens(0)._2 }else{ 0L }
+    val babelTokensDifference = recreatedBox.tokens(0)._2 - babelTokensBefore
+    val exchangeOK = babelTokensDifference * SELF.R5[Long].get >= nanoErgsDifference
+
+    sigmaProp(
+        babelFeeBoxCreator ||
+        (
+            babelFeeBoxRecreated && exchangeOK && (nanoErgsDifference >= 0)
+        )
+    ) 
+}
```

Comment:
That's ok. Babel fee box is in reality just an order book order to sell ERG in exchange for specified tokens at specified price. The babel fee box creator doesn't care about how these ERGs are spent - he earned revenue by selling these ERGs for a higher price than the ordinary market price (see text about financial incentives in this EIP). It's fully up to him to monitor whether he still wants to sell ERGs for the price specified in the R5, in case he no longer wants doing so, he can withdraw all funds from his babel fee box (and recreate it with new price in R5 in case he just wanted to update the price).